### PR TITLE
feat: HLS support for web

### PR DESCRIPTION
**Pull request by wanjm**

### CODE REVIEW

**Comment by alexmercerind** [comment]
Thanks!

**Comment by alexmercerind** [comment]
Hi!

I'm fine with taking this ahead, there are a few issues that I'm noticing. I will try to land it soon.

Thanks!

**Comment by wanjm** [comment]
ok, thanks.

File: media_kit/assets/hls1.4.10.js
Comment (alexmercerind) [inline media_kit/assets/hls1.4.10.js:1]: May we please place it as:

assets/web/hls.js

File: media_kit/lib/src/player/web/player/hls.dart
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/hls.dart:1]: - Please place this file inside lib/src/player/web/utils.
- Please format all Dart source-code files with `dart format`. We're following this practice.

FYI: You can configure Visual Studio Code with "Format on Save".

File: media_kit/lib/src/player/web/player/real.dart [hunk 1]
```diff
@@ -9,6 +9,7 @@ import 'dart:js' as js;
 import 'dart:typed_data';
 import 'dart:collection';
 import 'dart:html' as html;
+import 'package:media_kit/src/player/web/player/hls.dart';
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:12]: NIT: Please move this import below.

File: media_kit/lib/src/player/web/player/real.dart [hunk 2]
```diff
@@ -27,8 +28,19 @@ import 'package:media_kit/src/models/audio_params.dart';
 import 'package:media_kit/src/models/video_params.dart';
 import 'package:media_kit/src/models/playlist_mode.dart';
 
+
+void _loadHlsJs(){
+  final script = html.ScriptElement()
+    ..type = 'text/javascript'
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:34]: nit: sort assignment alphabetically except `src`.

File: media_kit/lib/src/player/web/player/real.dart [hunk 3]
```diff
@@ -27,8 +28,19 @@ import 'package:media_kit/src/models/audio_params.dart';
...
+    ..charset = 'utf-8'
+    ..async = true
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:36]: Does this cause any race condition?

File: media_kit/lib/src/player/web/player/real.dart [hunk 4]
```diff
@@ -368,6 +380,27 @@ class WebPlayer extends PlatformPlayer {
     }
   }
 
+  void loadResource(src) {
+    try {
+      if (src.contains(".m3u8") == false ||
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:385]: Please use `'` (single quote) strings like everywhere else in the source-code.

File: media_kit/lib/src/player/web/player/real.dart [hunk 5]
```diff
@@ -418,15 +451,7 @@ class WebPlayer extends PlatformPlayer {
...
-      try {
-        element.src = _playlist[_index].uri;
-      } catch (exception) {
-        // PlayerStream.error
-        final e = exception as html.DomException;
-        if (!errorController.isClosed) {
-          errorController.add(e.message ?? '');
-        }
-      }
+      loadResource(_playlist[_index].uri);
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:454]: Have you made sure to replace all `.src =` assignments?

File: media_kit/pubspec.yaml [hunk 6]
```diff
@@ -55,3 +56,6 @@ ffigen:
     entry-points:
       - headers/client.h
   dart-bool: true
+flutter:
```
Comment (alexmercerind) [inline media_kit/pubspec.yaml:59]: package:media_kit is a Dart (not Flutter) package. Does this work as intended?

NIT: Add a new line above it.

File: media_kit/lib/src/player/web/player/real.dart [hunk 7]
```diff
@@ -368,6 +380,27 @@ class WebPlayer extends PlatformPlayer {
     }
   }
 
+  void loadResource(src) {
+    try {
+      if (src.contains(".m3u8") == false ||
+          element.canPlayType('application/vnd.apple.mpegurl') != "") {
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:386]: Please use ' (single quote) strings like everywhere else in the source-code.

File: media_kit/lib/src/player/web/player/real.dart [hunk 8]
```diff
@@ -368,6 +380,27 @@ class WebPlayer extends PlatformPlayer {
     }
   }
 
+  void loadResource(src) {
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:383]: Is checking for .m3u8 in the URL sufficient? Are there any other checks we must have?

Maybe package:video_player_web_hls can serve as an example:

https://github.com/balvinderz/video_player_web_hls/blob/main/lib/src/video_player.dart#L319-L359

**Review by alexmercerind — CHANGES_REQUESTED** [review]
I'm going to request a few changes, in case you decide to work.

Though... I will try to research & fix these before landing.

<hr>

**Comment by alexmercerind** [comment]
Further Note: We follow [Convential Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification in our commit messages.

File: media_kit/assets/hls1.4.10.js
Comment (wanjm) [reply media_kit/assets/hls1.4.10.js:1]: I will move the directory and keep the name as hls1.4.10.js

File: media_kit/lib/src/player/web/player/real.dart (hunk 2 above)
Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart:34]: final script = html.ScriptElement()
    ..async = true
    ..charset = 'utf-8'
    ..type = 'text/javascript'
    ..src = 'assets/packages/media_kit/assets/hls1.4.10.js';

Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart:34]: final script = html.ScriptElement()
    ..async = true
    ..charset = 'utf-8'
    ..type = 'text/javascript'
    ..src = 'assets/packages/media_kit/assets/web/hls1.4.10.js';

File: media_kit/lib/src/player/web/player/real.dart (hunk 3 above)
Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart:36]: no

File: media_kit/lib/src/player/web/player/real.dart (hunk 8 above)
Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart:383]: > Is checking for .m3u8 in the URL sufficient? Are there any other checks we must have?
> 
> Maybe package:video_player_web_hls can serve as an example:
> 
> https://github.com/balvinderz/video_player_web_hls/blob/main/lib/src/video_player.dart#L319-L359

current i have no better idea. and the url can't be opened;

File: media_kit/lib/src/player/web/player/real.dart (hunk 5 above)
Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart]: > Have you made sure to replace all `.src =` assignments?

all are replaced;

File: media_kit/lib/src/player/web/player/real.dart (hunk 8 above)
Comment (alexmercerind) [reply media_kit/lib/src/player/web/player/real.dart]: I updated the URL in-case you decide to have a look.

Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart]: > I updated the URL in-case you decide to have a look.

found it

Comment (wanjm) [reply media_kit/lib/src/player/web/player/real.dart]: code added, but developer can't pass headers currently;

**Comment by alexmercerind** [comment]
Hi! I will land this PR soon, just a bit engaged.

**Comment by wanjm** [comment]
> Hi! I will land this PR soon, just a bit engaged.

ok

File: media_kit/test/src/player/player_test.dart [hunk 9]
```diff
@@ -3515,6 +3521,45 @@ Simply for <u>everyone</u>
     skip: kSkipFlakyTests,
     timeout: Timeout(const Duration(minutes: 2)),
   );
+  test(
+    'player-web-player-hls',
```
Comment (alexmercerind) [inline media_kit/test/src/player/player_test.dart:3525]: Added a unit-test.

File: media_kit/lib/src/player/web/player/real.dart [hunk 10]
```diff
@@ -302,6 +303,7 @@ class WebPlayer extends PlatformPlayer {
           }
         });
       });
+      await HLS.ensureInitialized(hls: test ? HLS.kHLSCDN : null);
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/player/real.dart:306]: Resolved possible race between HLS.js <script> load & `Player` creation.

File: media_kit/lib/src/player/web/utils/hls.dart [hunk 11]
```diff
@@ -0,0 +1,92 @@
+/// This file is a part of media_kit (https://github.com/media-kit/media-kit).
+///
+/// Copyright © 2021 & onwards, WanJiMi.
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/utils/hls.dart:3]: I added your name to this source file, in-case any change is required, raise a PR in future.

File: media_kit/lib/src/player/web/utils/hls.dart [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+          }
+        });
+        script.onError.listen((_) {
+          if (!completer.isCompleted) {
+            completer.completeError(Exception('Failed to load HLS.js'));
+          }
+        });
+
+        html.HeadElement? head = html.document.head;
+        if (head == null) {
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/utils/hls.dart:47]: Handled (possibly) hypothetical case where <head> tag is not located.

File: media_kit/lib/src/player/web/utils/hls.dart [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+/// Adds [HLS.js](https://github.com/video-dev/hls.js/) to the HTML document using [html.ScriptElement].
+///
+/// {@endtemplate}
+abstract class HLS {
+  static Future<void> ensureInitialized({String? hls}) {
+    return _lock.synchronized(() async {
+      if (_initialized) {
+        return;
+      }
+      final completer = Completer();
```
Comment (alexmercerind) [inline media_kit/lib/src/player/web/utils/hls.dart:27]: Wait until the HLS.js <script> is loaded successfully in lifecycle.

**Review by alexmercerind — APPROVED** [review]
I decided to remove the M3U8 check, it was not working & most importantly, it added additional delay for normal sources. I also noticed that all the test streams at HLS.js have "m3u8" in them:

https://github.com/video-dev/hls.js/blob/master/tests/test-streams.js

**Comment by alexmercerind** [comment]
Thanks :)

**Pull request merged**